### PR TITLE
Improve Condition Group in Advance Throttle Polcies

### PR DESCRIPTION
**Pull request by tharindu1st**
Previous Condition Validation happened on Gateway for every request with Every Condition From new Implementation GW receives the evaluated conditions from Traffic Manager.

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [x] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [x] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java [hunk 1]
```diff
@@ -170,6 +174,31 @@ private boolean isHeaderPresent(MessageContext messageContext, ConditionDTO cond
...
+                }
+                Pattern pattern = Pattern.compile(headerEntry.getValue());
+                Matcher matcher = pattern.matcher(value);
+                status = status && matcher.find();
+            }
+        }
+
+        if (condition.isInvert()){
+            return !status;
+        }else{
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java]: formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java [hunk 2]
```diff
@@ -170,6 +174,31 @@ private boolean isHeaderPresent(MessageContext messageContext, ConditionDTO cond
...
+                    status =  false;
+                    break;
+                }
+                Pattern pattern = Pattern.compile(headerEntry.getValue());
+                Matcher matcher = pattern.matcher(value);
+                status = status && matcher.find();
+            }
+        }
+
+        if (condition.isInvert()){
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java]: looks like code needs to be formatted noticed missing spaces in several places

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java [hunk 3]
```diff
@@ -170,6 +174,31 @@ private boolean isHeaderPresent(MessageContext messageContext, ConditionDTO cond
...
     }
 
+
+    private boolean isHeaderPresent(MessageContext messageContext, ConditionDto.HeaderConditions condition) {
+        TreeMap<String, String> transportHeaderMap = (TreeMap<String, String>) messageContext
+                .getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+        boolean status = true;
+        for (Map.Entry<String,String> headerEntry : condition.getValues().entrySet()){
+            if(transportHeaderMap != null){
+                String value = transportHeaderMap.get(headerEntry.getKey());
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java]: might use string utlls and there is a formatting issue next line

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java [hunk 4]
```diff
@@ -170,6 +174,31 @@ private boolean isHeaderPresent(MessageContext messageContext, ConditionDTO cond
...
+                if(value == null){
+                    status =  false;
+                    break;
+                }
+                Pattern pattern = Pattern.compile(headerEntry.getValue());
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java]: what happen value is null?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java [hunk 5]
```diff
@@ -184,6 +213,51 @@ private boolean isJWTClaimPresent(AuthenticationContext authenticationContext, C
...
+        for (Map.Entry<String, String> jwtClaim : condition.getValues().entrySet()) {
+            Object value = assertions.get(jwtClaim.getKey());
+            if (value == null) {
+                status = false;
+                break;
+            } else if (value instanceof String) {
+                String valueString = (String) value;
+                status = status && valueString.matches(jwtClaim.getValue());
+            } else {
+                status = status && false;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java]: cant we set status to false here? && means status is false

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java [hunk 6]
```diff
@@ -215,4 +289,106 @@ private boolean isWithinIP(MessageContext messageContext, ConditionDTO condition
...
+                    }
+                } else {
+                    condition = "default";
+                }
+            }
+        }
+        return condition;
+    }
+
+    private boolean isThrottledWithinCondition(MessageContext axis2MessageContext, AuthenticationContext authContext,
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java]: are we duplicating the code whih already available?

**Comment by harsha89** [comment]
Also please add more debug log statements.

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/policy_templates/throttle_policy_template_resource.xml [hunk 7]
```diff
@@ -15,19 +15,19 @@ ${policy.getTenantDomain()}_resource_${policy.getPolicyName()}_${pipeline}##
 @Import('org.wso2.throttle.processed.request.stream:1.0.0')
```
Comment (harsha89) [inline features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/policy_templates/throttle_policy_template_resource.xml]: do we require a migration?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java [hunk 8]
```diff
@@ -215,4 +289,106 @@ private boolean isWithinIP(MessageContext messageContext, ConditionDTO condition
...
+                                               List<ConditionDto> conditionDtoList) {
+        ThrottleProperties throttleProperties = ServiceReferenceHolder.getInstance().getThrottleProperties();
+        boolean status = true;
+        for (ConditionDto condition : conditionDtoList) {
+            if (condition.getIpCondition() != null) {
+                if (!isMatchingIP(axis2MessageContext, condition.getIpCondition())) {
+                    status &= false;
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java]: let's expand this for readability

Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleConditionEvaluator.java]: we can simply return false.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java [hunk 9]
```diff
@@ -270,33 +271,67 @@ private boolean doRoleBasedAccessThrottlingWithCEP(MessageContext synCtx, Config
                         // going to check it only if the API/Resource is throttled out.
                         if (getThrottleDataHolder().isAPIThrottled
                                 (resourceLevelThrottleKey)) {
+                            if (getThrottleDataHolder().isConditionsAvailable(resourceLevelThrottleKey)) {
+                                Map<String, List<ConditionDto>> conditionDtoMap = getThrottleDataHolder()
+                                        .getConditionDtoMap(resourceLevelThrottleKey);
+                                if (log.isDebugEnabled()) {
+                                    log.debug("conditions available" + conditionDtoMap.size());
+                                }
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java]: let's remove if check

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/policy_templates/throttle_policy_template_resource.xml [hunk 10]
```diff
@@ -15,19 +15,19 @@ ${policy.getTenantDomain()}_resource_${policy.getPolicyName()}_${pipeline}##
...
 define stream RequestStream (messageID string, appKey string, appTier string, subscriptionKey string, apiKey string, apiTier string, subscriptionTier string, resourceKey string, resourceTier string, userId string,  apiContext string, apiVersion string, appTenant string, apiTenant string, appId string, apiName string, propertiesMap string);
 
-@Export('org.wso2.throttle.globalThrottle.stream:1.0.0')
-define stream GlobalThrottleStream (throttleKey string, isThrottled bool, expiryTimeStamp long);
+@Export('org.wso2.throttle.globalThrottle.stream:1.1.0')
+define stream GlobalThrottleStream (throttleKey string, isThrottled bool, expiryTimeStamp long,evaluatedConditions string);
```
Comment (bhathiya) [inline features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/policy_templates/throttle_policy_template_resource.xml]: a space is missing.

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/policy_templates/throttle_policy_template_resource.xml (hunk 7 above)
Comment (tharindu1st) [reply features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/policy_templates/throttle_policy_template_resource.xml]: no

**Pull request merged**